pkg/controller/servicediscovery/privatednsnamespace/hooks.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/servicediscovery/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+func (e *external) observeImpl(ctx context.Context, cpm cpresource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	ret := managed.ExternalObservation{}
+	cr, ok := cpm.(*svcapitypes.PrivateDNSNamespace)
+	if !ok {
+		return ret, errors.New(errUnexpectedObject)
+	}
+	externalName := meta.GetExternalName(cr)
+	if externalName == "" {
+		return ret, nil
+	}
+
+	if cr.Status.AtProvider.OperationID == nil || *cr.Status.AtProvider.OperationID == "" {
+		return ret, nil
+	}
+	opInput := &svcsdk.GetOperationInput{
+		OperationId: cr.Status.AtProvider.OperationID,
+	}
+
+	opReqResp, err := e.client.GetOperationWithContext(ctx, opInput)
+	if err != nil {
+		return ret, errors.Wrap(err, "get-operation failed")
+	}
+
+	if *opReqResp.Operation.Status == "PENDING" || *opReqResp.Operation.Status == "SUBMITTED" {
+		ret.ResourceExists = true
+		return ret, nil
+	}
+
+	if *opReqResp.Operation.Status == "FAIL" {
+		if cr.GetCondition(xpv1.TypeReady).Reason != xpv1.ReasonDeleting {
+			ret.ResourceExists = true
+		}
+		cr.Status.SetConditions(xpv1.Unavailable().WithMessage(*opReqResp.Operation.ErrorMessage))
+		return ret, nil
+	}
+
+	namespaceID, ok := opReqResp.Operation.Targets["NAMESPACE"]
+	if !ok {
+		ret.ResourceExists = true
+		return ret, errors.New(errDescribe)
+	}
+
+	nsInput := &svcsdk.GetNamespaceInput{
+		Id: namespaceID,
+	}
+	nsReqResp, err := e.client.GetNamespaceWithContext(ctx, nsInput)
+	if err != nil {
+		if ActualIsNotFound(err) || IsDuplicateRequest(err) {
+			cr.Status.SetConditions(xpv1.Unavailable())
+		}
+		return ret, awsclients.Wrap(cpresource.IgnoreAny(err, ActualIsNotFound, IsDuplicateRequest), "get-namespace failed")
+	}
+	ret.ResourceExists = true
+
+	cr.Status.SetConditions(xpv1.Available())
+	if externalName != *namespaceID {
+		meta.SetExternalName(cr, *namespaceID)
+		ret.ResourceLateInitialized = true
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	lateInitialize(&cr.Spec.ForProvider, nsReqResp)
+	ret.ResourceLateInitialized = ret.ResourceLateInitialized || !cmp.Equal(&cr.Spec.ForProvider, currentSpec)
+
+	upToDate, err := e.isUpToDate(cr, nsReqResp)
+	if err != nil {
+		return ret, errors.Wrap(err, "isUpToDate check failed")
+	}
+
+	ret.ResourceUpToDate = upToDate
+	return ret, nil
+}
+
+func (e *external) updateImpl(ctx context.Context, mg cpresource.Managed) (managed.ExternalUpdate, error) {
+	// Update always returns an error, since the servicediscovery API does not allow updating namespaces
+	return managed.ExternalUpdate{}, nil
```

Comment:
In contrast to the comment `updateImpl` does not return an error

Reply:
This function was initially intended to always return an error but has since been reduced to this state. The function has been removed, as it double the nopUpdate function.